### PR TITLE
Update `tools` functions to work with n-dimensional fields

### PR DESCRIPTION
**Pull request by swapneelap**

### CODE REVIEW

File: discretisedfield/tests/test_tools.py [hunk 1]
```diff
@@ -64,56 +64,96 @@ def test_topological_charge(method):
...
-    p1 = (0, 0, 0)
-    p2 = (10, 10, 10)
-    cell = (2, 2, 2)
+@pytest.mark.parametrize("ndim", [1, 2, 3, 4])
+@pytest.mark.parametrize("nvdim", [1, 2, 3, 4])
+def test_emergent_magnetic_field(ndim, nvdim):
+    p1 = (0,) * ndim
+    p2 = (10,) * ndim
+    cell = (2,) * ndim
+    value = (0,) * nvdim
```
Comment (samjrholt) [inline discretisedfield/tests/test_tools.py:73]: If you are initialising with zeros then I don't think this is needed as the default value is zero anyway. As long as we have `nvdim=nvdim` when initalising.

File: discretisedfield/tests/test_tools.py [hunk 2]
```diff
@@ -64,56 +64,96 @@ def test_topological_charge(method):
...
-    with pytest.raises(ValueError):
-        dft.neigbouring_cell_angle(scalar_field, direction="x")
-
-    with pytest.raises(ValueError):
-        dft.neigbouring_cell_angle(field, direction="l")
-
-    with pytest.raises(ValueError):
-        dft.neigbouring_cell_angle(field, direction="x", units="wrong")
+    # Check for a value of angle
+    arr_x_coord = mesh.coordinate_field().array[..., 0, np.newaxis].copy()
```
Comment (samjrholt) [inline discretisedfield/tests/test_tools.py:117]:
```suggestion
    arr_x_coord = mesh.coordinate_field().g1.array
```
Could we do?

File: discretisedfield/tests/test_tools.py [hunk 3]
```diff
@@ -64,56 +64,96 @@ def test_topological_charge(method):
...
+    arr = np.concatenate(
+        [
+            np.full((*mesh.n, 1), 0),
+            np.cos(arr_x_coord * np.pi / 20),
+            np.sin(arr_x_coord * np.pi / 20),
+        ],
+        axis=-1,
+    )
+    fied_piby2 = df.Field(mesh, nvdim=3, value=arr)
+    assert np.isclose(
```
Comment (samjrholt) [inline discretisedfield/tests/test_tools.py:127]: Nice example

**Review by samjrholt — APPROVED** [review]
Looks good, just a few minor comments

**Review by samjrholt — APPROVED** [review]

File: discretisedfield/tests/test_tools.py (hunk 2 above)
Comment (swapneelap) [reply discretisedfield/tests/test_tools.py:116]: This unfortunately will ask for the `g1` component of the vector which does not exist. `g1` is a geometric dimension here.

File: discretisedfield/tests/test_tools.py (hunk 1 above)
Comment (swapneelap) [reply discretisedfield/tests/test_tools.py]: Nicely spotted! 😃

File: discretisedfield/tests/test_tools.py [hunk 4]
```diff
@@ -64,56 +64,96 @@ def test_topological_charge(method):
...
+        g1, *_ = point
+        if g1 < 50:
+            return (0, 0, 1)
+        else:
+            return (0, 0, -1)
+
+    field.update_field_values(val)
+
+    # Here the mean is 2*pi/10 because 2 cells will have pi values
+    assert np.isclose(dft.max_neighbouring_cell_angle(field).mean(), 2 * np.pi / 10)
```
[the remark this answers is not in the document]
Comment (swapneelap) [reply discretisedfield/tests/test_tools.py:155]: I was inspired by your examples earlier 😃

File: discretisedfield/tests/test_tools.py (hunk 2 above)
Comment (swapneelap) [reply discretisedfield/tests/test_tools.py:116]: Yes, it works but not for `vdims=1` and `vdims=4` 😅 . I have a work-around though, I'll push a commit and tag you again.

Comment (samjrholt) [reply discretisedfield/tests/test_tools.py]: @swapneelap sorry for the poor explication, I have just committed what I was meaning. Take a look and feel free to remove the commit if you want but it works for all `ndim` and the tests pass but I may be missing something.

Comment (swapneelap) [reply discretisedfield/tests/test_tools.py]: I had exactly the same thing in mind 😄 . Thank you for the commit!

**Pull request merged**